Cookbook/Technical-Documents/Current-guide.md

```diff
@@ -0,0 +1,40 @@
+# `Current` Guide - how to use it in Babylon iOS Project
+
+### Where can I find it?
+- `Current` is the global instance of the `World` type which is located in the `BabylonDependencies.framework`.
+Its purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns). 
+For more information please read the [How Control the World](/Cookbook/Proposals/ControlTheWorld.md) proposal.
+
+### Beyond the Proposal
+All the properties in `Current` must have their initial values provided at initialisation time. 
+But that's not the reality for all of them. These `async` properties have their initial value provided in the `SharedAppDelegate` through the method `setupCurrent`. Only after that, the whole world is ready to be used.
+
+## `Do`s and `Don't`s
+
+#### Do
+- access `Current` **only** from topmost `AppDelegate` until `Builder`s layer
+- pass the required dependencies down the chain;
+- create a customized version to fit your specific needs **for tests** and set it in the test `setUp` method. Please reset it on the `tearDown` method 
+```swift
+override func setUp() {
+    let utc = TimeZone(identifier: "UTC")!
+    let locale = Locale(identifier: "en_GB")
+    Current = World(autoupdatingTimeZone: utc, autoupdatingLocale: locale)
+}
+
+override func tearDown() {
+    Current = .production // which returns a new `World()`
+}
+```
+#### Don't
+- access `Current` from any other layer other than `Builder`. All the other layers (`ViewModel`, `FlowController`, `ViewController`, `Model`, `BusinessController`) continue to receive their dependencies through injection when created by the builder, preferably at initialisation time;
+- inject `current` as a dependency;
+- try to mutate `Current` (excluding Tests). In `Release` configuration it's a constant `let`;
```

Comment:
Should we remove the `Current.changeTheme` method from `World` then?

Is there a difference in that sentence between "mutating the content of `Current`" and "replace the `Current` instance with another one"? iow, are we both prohibited from doing `Current = someNewValue` but also `Current.changeWhatever` – assuming `changeWhatever` mutates some properties of the `Current` instance?

I think this point deserves clarification in this sentence to disambiguate between the two cases.

Reply:
Perfect point @AliSoftware 

There is a bunch of `set` methods:
`setBiometricAuth`, `setNetworkDependencies`, `setEnvironment`, `setAppConfiguration` and also as you said the `changeTheme`. 

These methods change **values** inside Current's private mutable properties that are consumed by its public immutable computed ones. Basically value semantics over mutability.

Answering your question: both are prohibited. The only entity allowed to run these `set` methods is the SharedAppDelegate.

I wonder if changing the:
> (excluding Tests)

to 

> (excluding Test targets and `SharedAppDelegate.setupCurrent` methods)

Would it disambiguate that point?

> - try to mutate `Current` (excluding Test targets and `SharedAppDelegate.setupCurrent` methods). In `Release` configuration it's a constant `let`;

Reply:
I think it would be useful to specify `excluding test targets and setupCurrent` indeed, but I don't think that would be enough to clarify that both "Replacing the Current instance altogether" and "mutating the content of Current" are prohibited elsewhere.

I mean if you look at the code you can eventually see that `World` is a `struct` so has value semantics and thus when it's `let` (in Release) you won't be able to mutate even its content, so you can _guess_ and _conclude_ that indeed both should not be allowed anyway, but making it explicit in such a document dedicated to `Current` still feels useful to me.

```suggestion
- Mutate `Current`'s properties – or re-affect it to another instance – after `SharedAppDelegate.setupCurrent()`. You are only allowed to mutate `Current` in tests, not in app. In `Release` configuration it's a constant `let` anyway to prevent mutation;
```